dnschecker.go

```diff
@@ -0,0 +1,120 @@
+package checkup
+
+import (
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/miekg/dns"
+)
+
+// DNSChecker implements a Checker for TCP endpoints.
+type DNSChecker struct {
+	// EndpointName is the name of the endpoint.
+	EndpointName string `json:"endpointname"`
+	// This is the name of the DNS server you are testing.
+	EndpointURL string `json:"endpoint_url"`
+	// This is the fqdn of the target server to query the DNS server for.
+	HostnameFQDN string `json:"hostname_fqdn,omitempty"`
+	// Timeout is the maximum time to wait for a
+	// TCP connection to be established.
+	Timeout time.Duration `json:"timeout,omitempty"`
+	// ThresholdRTT is the maximum round trip time to
+	// allow for a healthy endpoint. If non-zero and a
+	// request takes longer than ThresholdRTT, the
+	// endpoint will be considered unhealthy. Note that
+	// this duration includes any in-between network
+	// latency.
+	ThresholdRTT time.Duration `json:"threshold_rtt,omitempty"`
+	// Attempts is how many requests the client will
+	// make to the endpoint in a single check.
+	Attempts int `json:"attempts,omitempty"`
+}
+
+// Check performs checks using c according to its configuration.
+// An error is only returned if there is a configuration error.
+func (c DNSChecker) Check() (Result, error) {
+	if c.Attempts < 1 {
+		c.Attempts = 1
+	}
+
+	result := Result{Title: c.EndpointName, Endpoint: c.EndpointURL, Timestamp: Timestamp()}
+	result.Times = c.doChecks()
+
+	return c.conclude(result), nil
+}
+
+// doChecks executes and returns each attempt.
+func (c DNSChecker) doChecks() Attempts {
+	var err error
+	var conn net.Conn
+
+	timeout := c.Timeout
+	if timeout == 0 {
+		timeout = 1 * time.Second
+	}
+	_ = timeout
```

Comment:
Metalinter coughed on it, I didn’t mean to leave it there, let me resolve.
> On Aug 19, 2016, at 1:26 PM, Matt Holt <notifications@github.com> wrote:
> 
> In dnschecker.go <https://github.com/sourcegraph/checkup/pull/12#discussion_r75529770>:
> 
> > +	result := Result{Title: c.EndpointName, Endpoint: c.EndpointURL, Timestamp: Timestamp()}
> > +	result.Times = c.doChecks()
> > +
> > +	return c.conclude(result), nil
> > +}
> > +
> > +// doChecks executes and returns each attempt.
> > +func (c DNSChecker) doChecks() Attempts {
> > +	var err error
> > +	var conn net.Conn
> > +
> > +	timeout := c.Timeout
> > +	if timeout == 0 {
> > +		timeout = 1 * time.Second
> > +	}
> > +	_ = timeout
> What's this?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/sourcegraph/checkup/pull/12/files/3675fd806e5e2eadcecbe88e643ca87b20b70336#r75529770>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AT4YpXQ2sj7uhBSzqS4Tr5IZT6La-jkbks5qhfVdgaJpZM4Jij2O>.
>